src/Control/Distributed/Closure/Class.hs

```diff
@@ -0,0 +1,81 @@
+-- | 'Closure' is not a functor, since we cannot map arbitrary functions over
+-- it. But it sure looks like one, and an applicative one at that. What we can
+-- do is map /static pointers to/ arbitrary functions over it (or in general,
+-- closures). 'Closure' is not just an applicative functor, it's also a monad,
+-- as well as a comonad, if again we limit the function space to those functions
+-- that can be statically pointed to.
+--
+-- In fact an entire hierarchy of classes mirroring the standard classes can be
+-- defined, where nearly the only difference lies in the fact that higher-order
+-- arguments must be a proof of /static/-ness (i.e. a 'Closure'). The other
+-- difference is that composing static values requires a proof of typeability,
+-- so we carry those around ('Typeable' constraints).
+--
+-- This module defines just such a class hierarchy in the category of static
+-- functions (aka values of type @'Closure' (a -> b)@).
+
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE StaticPointers #-}
+{-# LANGUAGE UndecidableInstances #-} -- For StaticMonad instance.
+
+module Control.Distributed.Closure.Class
+  ( Static(..)
+  , StaticFunctor(..)
+  , StaticApply(..)
+  , StaticBind(..)
+  , StaticApplicative(..)
+  , StaticMonad
+  , staticReturn
+  , StaticComonad(..)
+  ) where
+
+import Control.Distributed.Closure
+import Data.Typeable (Typeable)
+
+class Typeable f => StaticFunctor f where
+  staticMap :: (Typeable a, Typeable b) => Closure (a -> b) -> f a -> f b
+
+class StaticFunctor f => StaticApply f where
+  staticApply :: (Typeable a, Typeable b) => f (a -> b) -> f a -> f b
+
+class StaticApply f => StaticApplicative f where
+  staticPure :: Typeable a => a -> f a
+
+class StaticApply m => StaticBind m where
+  staticBind :: (Typeable a, Typeable b) => m a -> Closure (a -> m b) -> m b
+  staticBind m k = staticJoin (staticMap k m)
+
+  staticJoin :: Typeable a => m (m a) -> m a
+  staticJoin m = m `staticBind` static id
+
+  {-# MINIMAL staticBind | staticJoin #-}
+
+class (StaticApplicative m, StaticBind m) => StaticMonad m
+instance (StaticApplicative m, StaticBind m) => StaticMonad m
+
+staticReturn :: (StaticApplicative m, Typeable a) => a -> m a
```

Comment:
The idea here isn't to rewrite Haskell history, except where we have to (in particular the semigroupoid type classes are a necessity for us). I did move `staticReturn` out of being a class method. But principle of least surprise and all, I think it makes sense to retain the symmetry with standard type classes and have both names available.